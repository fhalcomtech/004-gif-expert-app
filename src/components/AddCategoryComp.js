import React, { useState } from 'react'

export const AddCategoryComp = ({setCategories}) => {
    const [category, setCategory] = useState('');
    const [timer, setTimer] = useState(0);
    const [msg, setMsg] = useState('');
    const handlerChangeCategory = (e) => setCategory(e.target.value); 
    const handlerFormSubmit = (e) => {
        e.preventDefault();
        if(category.length<1) return;

        setCategories(curr => {
            if(timer){
                setTimer(tm => {
                    clearTimeout(tm);
                    return 0;
                });
            }
            if(curr.includes(category))
            {
                setMsg('La categoria ya existe')
                return curr;
            }
            else
            {
                setMsg('La categoria fue Agregada con exito')
                return ([category, ...curr]);
            }
        });
        setCategory('');
        setTimer(tm=>setTimeout(()=>{setMsg('')},5000));
        
    }

  return (
      <form onSubmit={handlerFormSubmit}>
          <h2>Gif Expert App</h2> 
          <h5>{msg}</h5>
          <input type='text' value={category} onChange={handlerChangeCategory}/>
          <button type='submit'>Save</button>
      </form>
  )
}
