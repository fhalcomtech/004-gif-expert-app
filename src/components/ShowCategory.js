import React from 'react'
import PropTypes from 'prop-types';

export const ShowCategory = ({ url, title}) => {
  return (
    <div className='card'>
        <div className='card-img'>
            <img src={url} alt={title}/>
        </div>    
        <div className='card-name'>
            {title.slice(0,10)}
        </div>
    </div>
  )
}

ShowCategory.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
}
