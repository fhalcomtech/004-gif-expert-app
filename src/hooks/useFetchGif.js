import { useEffect, useState } from "react";
import { getGifHelper } from "../helpers/fetchGitHelper";

export const useFetchGif = (category) => {
  const[gifs, setGifs] = useState({loading: true, data:[]});

  useEffect(()=>
  {
      const gifsDataLst = getGifHelper(category);
      gifsDataLst.then(data => {setGifs({loading:false, data:data})});
    },[category])
  
  return gifs;
}
