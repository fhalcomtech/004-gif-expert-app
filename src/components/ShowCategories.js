import { useFetchGif } from '../hooks/useFetchGif';
import { ShowCategory } from './ShowCategory';

export const ShowCategories = ({category}) => {
   const {loading, data:images} = useFetchGif(category);
  return (
    <div className='category-container'>
        <h4 className='category-name'>
            {loading?'Loading...':category} 
        </h4>
        <div className='category-images'>
            {images.map(({id, url, title}) => (<ShowCategory key={id} url={url} title={title}/>))}
        </div>
    </div>
  )
}
