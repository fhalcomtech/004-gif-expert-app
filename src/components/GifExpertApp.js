import React, { useState } from 'react';
import { AddCategoryComp } from './AddCategoryComp';
import { ShowCategories } from './ShowCategories';

const GifExpertApp = ({defaultCategories= []}) =>
{
  const [categ, setCateg] = useState(defaultCategories);
  return (
    <>
      <AddCategoryComp setCategories={setCateg} />
      <hr/>
      <div>
         <ul>{categ.map(ct=><ShowCategories key={ct} category={ct} />)}</ul>
      </div>
    </>
  )

  
}

export default GifExpertApp;