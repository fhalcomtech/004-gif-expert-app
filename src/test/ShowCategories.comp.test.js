import {shallow} from "enzyme";
import {ShowCategories} from "../components/ShowCategories";
import {useFetchGif} from "../hooks/useFetchGif";
jest.mock("../hooks/useFetchGif");
describe('Tests ShowCategories Component', ()=>{
    const category = 'goku';

    test('Create ShowCategories component snapshot',()=>{
        useFetchGif.mockReturnValue({loading: true, data:[]})
        const wrapper = shallow(<ShowCategories category={category}/>)
        expect(wrapper).toMatchSnapshot();
    })


    test('Create ShowCategories component snapshot with fetch data',()=>{
        const nData = [{id:"xyz", url:"http://localhost/img/fhalcom.gif", title:"fhalcomX"},
            {id:"abc", url:"http://localhost/img/abc.jpg", title:"abc"}];
        useFetchGif.mockReturnValue({loading: false, data:nData})
        const wrapper = shallow(<ShowCategories category={category}/>)
        expect(wrapper).toMatchSnapshot();
    })
})