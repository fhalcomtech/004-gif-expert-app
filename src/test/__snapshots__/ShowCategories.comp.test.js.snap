// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests ShowCategories Component Create ShowCategories component snapshot 1`] = `
<div
  className="category-container"
>
  <h4
    className="category-name"
  >
    Loading...
  </h4>
  <div
    className="category-images"
  />
</div>
`;

exports[`Tests ShowCategories Component Create ShowCategories component snapshot with fetch data 1`] = `
<div
  className="category-container"
>
  <h4
    className="category-name"
  >
    goku
  </h4>
  <div
    className="category-images"
  >
    <ShowCategory
      key="xyz"
      title="fhalcomX"
      url="http://localhost/img/fhalcom.gif"
    />
    <ShowCategory
      key="abc"
      title="abc"
      url="http://localhost/img/abc.jpg"
    />
  </div>
</div>
`;
